standalone-metastore/src/main/java/org/apache/hadoop/hive/metastore/txn/TxnHandler.java

```diff
@@ -585,24 +586,11 @@ public OpenTxnsResponse openTxns(OpenTxnRequest rqst) throws MetaException {
         stmt = dbConn.createStatement();
 
         if (rqst.isSetReplPolicy()) {
-          List<String> inQueries = new ArrayList<>();
-          StringBuilder prefix = new StringBuilder();
-          StringBuilder suffix = new StringBuilder();
-
-          prefix.append("select RTM_TARGET_TXN_ID from REPL_TXN_MAP where ");
-          suffix.append(" and RTM_REPL_POLICY = " + quoteString(rqst.getReplPolicy()));
-
-          TxnUtils.buildQueryWithINClause(conf, inQueries, prefix, suffix, rqst.getReplSrcTxnIds(),
-                  "RTM_SRC_TXN_ID", false, false);
-
-          for (String query : inQueries) {
-            LOG.debug("Going to execute select <" + query + ">");
-            rs = stmt.executeQuery(query);
-            if (rs.next()) {
-              LOG.info("Transactions " + rqst.getReplSrcTxnIds().toString() +
-                      " are already present for repl policy " + rqst.getReplPolicy());
-              return new OpenTxnsResponse(new ArrayList<>());
-            }
+          List<Long> targetTxnIdList = getTargetTxnIdList(rqst.getReplPolicy(), rqst.getReplSrcTxnIds(), stmt);
+          if (!targetTxnIdList.isEmpty()) {
+            LOG.info("Target transactions " + targetTxnIdList.toString() + " is present for repl policy :" +
+              rqst.getReplPolicy() + " and Source transaction id : " + rqst.getReplSrcTxnIds().toString());
+            return new OpenTxnsResponse(new ArrayList<>());
```

Comment:
We claim openTxn is idempotent right? So, can we just return the target txn ids instead of empty list?

Reply:
done